### ISSUE TITLE
[Feature]: Can Milvus support custom vector distance calculation methods?

### ISSUE THREAD

**Issue by cheetahsec** [milvus-io/milvus#25701]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe.

In the product context, it is essential to retrieve file feature embedding distances, which involve several sets of calculations for embedding distances, including:

> Core structure of file features:

- X (entropy coefficient) + Y (distribution coefficient) + Z (content embedding vector with 256 dimensions)

> Method for calculating feature embedding distances (using objects A and B as examples):

- distance = mod_diff(A.X - B.X) * X_Scale + mod_diff(A.Y - B.Y) * Y_Scale + distance_func(A.Z, B.Z)
- Explanation of relevant parameters:
- mod_diff: Calculates the minimal distance between two points on a circle
- X_Scale: Weighted value of the X coefficient (business tuning)
- Y_Scale: Weighted value of the Y coefficient (business tuning)
- distance_func: Distance calculation function, similar to Hamming distance, adjusted and optimized based on business data

The current objective is to replace the original search framework used in the product with Milvus. However, the distance calculation only supports standard distance functions, making it challenging for the business to adapt the embedding expression.

### Describe the solution you'd like.

Does Milvus support custom embedding distance calculations? This would allow for better alignment with business requirements without the need to consider alternative embedding expression transformations.

### Describe an alternate solution.

The current approach entails:

1. Retrieving the embeddings for the Z component using Milvus, filtering the TopK results, and recalculating and sorting based on the business distance function. While this satisfies most requirements, some unique scenarios may result in missed outcomes.
2. Normalizing and transforming X, Y, and Z into floating-point vectors, multiplying them by their respective weighted coefficients, and converting them into final embeddings. Employing the L2 distance for retrieval may yield inconsistencies between the distance distribution scales and the original distance distribution, potentially causing false negatives and false positives in certain cases.

**Comment by cheetahsec** [comment]
Upon comparing project historical documents and the TLSH implementation paper, the implementations are essentially consistent. Therefore, if considering distance metrics that cover TLSH, the official paper at https://tlsh.org/papers.html can be consulted for more technical details. TLSH has extensive applications in the security domain and is currently an industry standard in open-source implementations. Its versatility in file similarity comparisons is also commendable. It is worth considering support for TLSH, akin to the currently implemented Superstructure and Substructure metrics for molecular structure similarity measurements.

**Comment by suizhihao** [comment]
So how to query? For example, there is a vector_index which IS 2048 dimensions, the first 128 are obtained by embedding col_1, the middle 512 dimensions are obtained by col_2, and the last part is obtained by col_3. Can I specify which part to query?

**Comment by cheetahsec** [comment]
> 

/ The stored data and the target of the query are hash values, hash value content contains multi-dimensional comparison results, so the distance calculation between hash values is non-standard, a certain part of the query alone is inaccurate, and your problem may not be exactly the same, if you mention the three vectors are completely unrelated, then there is no need to store together, if there is a relationship you can try to split into three queries, and finally calculate the score according to the weighting of demand

**Comment by xiaofan-luan** [comment]
> 

I think we don't really need to worry about hot it is calculated.
It's more like we implement a function template and load the user specified so

**Comment by cheetahsec** [comment]
Yes, after trying to modify knowhere to add a custom distance calculation formula, we think that a native plug-in mechanism should be provided to meet similar needs. In addition, there is confusion in the selection of specific index types. The TLSH distance should belong to the non-metric space. Is there only HNSW indexes support similar needs

**Comment by liliu-z** [comment]
> Yes, after trying to modify knowhere to add a custom distance calculation formula, we think that a native plug-in mechanism should be provided to meet similar needs. In addition, there is confusion in the selection of specific index types. The TLSH distance should belong to the non-metric space. Is there only HNSW indexes support similar needs

IMHO, yes. But you can try IVF-series, since it works in MIPS, which is still a non-metric problem. It depends on your datasets.

**Comment by cheetahsec** [comment]
> > 
> 
> I think we don't really need to worry about hot it is calculated. It's more like we implement a function template and load the user specified so

Is there a specific development plan and timeline for custom distance plugins?

**Comment by ramialt** [comment]
I'd like to chime in and say that one of the big reasons I chose Milvus was that it was the easiest one to modify to add a custom distance function for my use-case. This was because it already used popular libraries to do vector indexing.

In my case I already had written my custom distance function in C++ inside one of those libraries. So, from there I just had to add my custom distance as an option in a couple of Milvus files (I can post the list of the files edited if desired).

I think having this option would make Milvus look a lot more desirable comparable to other vector databases.
Compared to my handmade solution, having Milvus handle custom distances would allow me to easily update Milvus versions without having to manually redo all the changes I've made. My only concern is the potential performance drop of my C-defined distance function vs. the Milvus SDK-defined (likely Python) version of the distance function. But either way, I still think this feature would be very helpful.

**Comment by xiaofan-luan** [comment]
> I'd like to chime in and say that one of the big reasons I chose Milvus was that it was the easiest one to modify to add a custom distance function for my use-case. This was because it already used popular libraries to do vector indexing.
> 
> In my case I already had written my custom distance function in C++ inside one of those libraries. So, from there I just had to add my custom distance as an option in a couple of Milvus files (I can post the list of the files edited if desired).
> 
> I think having this option would make Milvus look a lot more desirable comparable to other vector databases. Compared to my handmade solution, having Milvus handle custom distances would allow me to easily update Milvus versions without having to manually redo all the changes I've made. My only concern is the potential performance drop of my C-defined distance function vs. the Milvus SDK-defined (likely Python) version of the distance function. But either way, I still think this feature would be very helpful.

Good job @ramialt 
It should be too hard to implement your own metrics but for the community and User defined code is a concern

**Comment by cheetahsec** [comment]
> I'd like to chime in and say that one of the big reasons I chose Milvus was that it was the easiest one to modify to add a custom distance function for my use-case. This was because it already used popular libraries to do vector indexing.我想插话一下，我选择 Milvus 的一大原因是它是最容易修改的，可以为我的用例添加自定义距离函数。这是因为它已经使用流行的库来进行向量索引。
> 
> In my case I already had written my custom distance function in C++ inside one of those libraries. So, from there I just had to add my custom distance as an option in a couple of Milvus files (I can post the list of the files edited if desired).就我而言，我已经在其中一个库中用 C++ 编写了自定义距离函数。因此，从那里我只需将自定义距离添加为几个 Milvus 文件中的选项（如果需要，我可以发布编辑的文件列表）。
> 
> I think having this option would make Milvus look a lot more desirable comparable to other vector databases.我认为有了这个选项会让 Milvus 看起来比其他矢量数据库更有吸引力。 Compared to my handmade solution, having Milvus handle custom distances would allow me to easily update Milvus versions without having to manually redo all the changes I've made. My only concern is the potential performance drop of my C-defined distance function vs. the Milvus SDK-defined (likely Python) version of the distance function. But either way, I still think this feature would be very helpful.与我手工制作的解决方案相比，让 Milvus 处理自定义距离可以让我轻松更新 Milvus 版本，而无需手动重做我所做的所有更改。我唯一担心的是我的 C 定义的距离函数与 Milvus SDK 定义的（可能是 Python）版本的距离函数相比，潜在的性能下降。但无论如何，我仍然认为这个功能会非常有帮助。

If it is convenient to publish, the list of modified and edited files will be a good reference case, and no related sharing has been seen yet. I have similar concerns about performance and may need to refer to SIMD acceleration. It is best to have an officially supported implementation mechanism for the custom distance, otherwise version synchronization will also be a problem.

**Comment by xiaofan-luan** [comment]
> > I'd like to chime in and say that one of the big reasons I chose Milvus was that it was the easiest one to modify to add a custom distance function for my use-case. This was because it already used popular libraries to do vector indexing.我想插话一下，我选择 Milvus 的一大原因是它是最容易修改的，可以为我的用例添加自定义距离函数。这是因为它已经使用流行的库来进行向量索引。
> > In my case I already had written my custom distance function in C++ inside one of those libraries. So, from there I just had to add my custom distance as an option in a couple of Milvus files (I can post the list of the files edited if desired).就我而言，我已经在其中一个库中用 C++ 编写了自定义距离函数。因此，从那里我只需将自定义距离添加为几个 Milvus 文件中的选项（如果需要，我可以发布编辑的文件列表）。
> > I think having this option would make Milvus look a lot more desirable comparable to other vector databases.我认为有了这个选项会让 Milvus 看起来比其他矢量数据库更有吸引力。 Compared to my handmade solution, having Milvus handle custom distances would allow me to easily update Milvus versions without having to manually redo all the changes I've made. My only concern is the potential performance drop of my C-defined distance function vs. the Milvus SDK-defined (likely Python) version of the distance function. But either way, I still think this feature would be very helpful.与我手工制作的解决方案相比，让 Milvus 处理自定义距离可以让我轻松更新 Milvus 版本，而无需手动重做我所做的所有更改。我唯一担心的是我的 C 定义的距离函数与 Milvus SDK 定义的（可能是 Python）版本的距离函数相比，潜在的性能下降。但无论如何，我仍然认为这个功能会非常有帮助。
> 
> If it is convenient to publish, the list of modified and edited files will be a good reference case, and no related sharing has been seen yet. I have similar concerns about performance and may need to refer to SIMD acceleration. It is best to have an officially supported implementation mechanism for the custom distance, otherwise version synchronization will also be a problem.

Even if we support User defined distance calculation, the SIMD optimization still have to be done by user or by the compiler. I guess this has to be dramatically slow than highly optimized L2 or IP

**Comment by cheetahsec** [comment]
> > > I'd like to chime in and say that one of the big reasons I chose Milvus was that it was the easiest one to modify to add a custom distance function for my use-case. This was because it already used popular libraries to do vector indexing.我想插话一下，我选择 Milvus 的一大原因是它是最容易修改的，可以为我的用例添加自定义距离函数。这是因为它已经使用流行的库来进行向量索引。
> > > In my case I already had written my custom distance function in C++ inside one of those libraries. So, from there I just had to add my custom distance as an option in a couple of Milvus files (I can post the list of the files edited if desired).就我而言，我已经在其中一个库中用 C++ 编写了自定义距离函数。因此，从那里我只需将自定义距离添加为几个 Milvus 文件中的选项（如果需要，我可以发布编辑的文件列表）。
> > > I think having this option would make Milvus look a lot more desirable comparable to other vector databases.我认为有了这个选项会让 Milvus 看起来比其他矢量数据库更有吸引力。 Compared to my handmade solution, having Milvus handle custom distances would allow me to easily update Milvus versions without having to manually redo all the changes I've made. My only concern is the potential performance drop of my C-defined distance function vs. the Milvus SDK-defined (likely Python) version of the distance function. But either way, I still think this feature would be very helpful.与我手工制作的解决方案相比，让 Milvus 处理自定义距离可以让我轻松更新 Milvus 版本，而无需手动重做我所做的所有更改。我唯一担心的是我的 C 定义的距离函数与 Milvus SDK 定义的（可能是 Python）版本的距离函数相比，潜在的性能下降。但无论如何，我仍然认为这个功能会非常有帮助。
> > 
> > 
> > If it is convenient to publish, the list of modified and edited files will be a good reference case, and no related sharing has been seen yet. I have similar concerns about performance and may need to refer to SIMD acceleration. It is best to have an officially supported implementation mechanism for the custom distance, otherwise version synchronization will also be a problem.如果方便发布的话，修改编辑的文件列表会是一个很好的参考案例，目前还没有看到相关的分享。我对性能也有类似的担忧，可能需要参考 SIMD 加速。自定义距离最好有官方支持的实现机制，否则版本同步也会成为问题。
> 
> Even if we support User defined distance calculation, the SIMD optimization still have to be done by user or by the compiler. I guess this has to be dramatically slow than highly optimized L2 or IP即使我们支持用户定义的距离计算，SIMD 优化仍然必须由用户或编译器完成。我猜这肯定比高度优化的 L2 或 IP 慢得多

Yes, but compared to modifying the entire project code by self, users only need to bear the performance of distance calculation, which is still a very attractive feature.

**Comment by ramialt** [comment]
As requested by @cheetahsec , the list of files I modified to add custom distance implementation is as follows:

Inside Milvus (used version 2.2.12)
* internal/core/src/segcore/SegcoreConfig.h
* internal/proxy/util.go
* internal/util/indexparamcheck/constraints.go
* internal/core/src/query/SearchBruteForce.cpp
* internal/core/src/query/deprecated/ValidationUtil.cpp
   * Not sure why I had to modify this one since it says deprecated, but I believe I got an error message from this file ¯\\\_(ツ)\_/¯

Inside Knowhere (used version 1.3.17)
* knowhere/index/vector_index/ConfAdaptor.cpp
* knowhere/index/vector_index/Index<NAME\>.cpp (per index)
* knowhere/index/vector_index/helpers/IndexParameter.h
* thirdparty/*  -- Files in whichever third-party lib used (faiss, hnswlib)

**Comment by cheetahsec** [comment]
> 

good job, thank you very much for your help.

**Comment by xiaofan-luan** [comment]
not the goal for 2.5, move to 2.x backlog

**Comment by xiaofan-luan** [comment]
python is kind like too slow. we will need to load .so or wasm.
